lampe/inference.py

```diff
@@ -532,6 +548,177 @@ def forward(self, theta: Tensor, x: Tensor) -> Tensor:
+
+    where
+
+    .. math::
+        \alpha(t) & = \exp \left( -\int_0^t \beta(u) \, \mathrm{d} u \right) \\
+        \beta(t) & = (\beta_\max - \beta_\min) \, t + \beta_\min .
+
+    Because the drift coefficient is affine, the perturbation kernel
+    :math:`p(\theta_t | \theta)` corresponding to this diffusion process is Gaussian
+    and takes the form :math:`\mathcal{N}(\sqrt{\alpha(t)} \, \theta, (1 - \alpha(t))^2 I)`,
+    which enables efficient training through the (rescaled) score-matching objective
+
+    .. math:: \arg\min_\phi \mathbb{E}_{p(\theta, x) p(t) p(\theta_t | \theta)}
+        \Big[ \big\|
+            s_\phi(\theta_t, x, t) -
+            (1 - \alpha(t)) \nabla_{\! \theta_t} \log p(\theta_t | \theta)
+        \big\|_2^2 \Big] .
+
+    For log-density computation and sampling, we solve the probability flow ODE
+
+    .. math:: \mathrm{d} \theta = \left[ -\frac{1}{2} \beta(t) \theta -
+        \frac{1}{2} \beta(t) (1 + \alpha(t)) s_\phi(\theta, x, t) \right] \, \mathrm{d} t
+
+    whose trajectories share the same marginal probability densities :math:`p(\theta_t | x)`
+    as the diffusion process.
+
+    References:
+        | Score-Based Generative Modeling through Stochastic Differential Equations (Song et al., 2021)
+        | https://arxiv.org/abs/2011.13456
+
+    Arguments:
+        theta_dim: The dimensionality :math:`D` of the parameter space.
+        x_dim: The dimensionality :math:`L` of the observation space.
+        t_dim: The dimensionality of the time embedding.
+        moments: The parameters moments :math:`\mu` and :math:`\sigma`. If provided,
+            the moments are used to standardize the parameters.
+        build: The network constructor (e.g. :class:`lampe.nn.ResMLP`).
+        kwargs: Keyword arguments passed to the constructor.
+    """
+
+    def __init__(
+        self,
+        theta_dim: int,
+        x_dim: int,
+        t_dim: int = 1,
+        moments: Tuple[Tensor, Tensor] = None,
+        build: Callable[[int, int], nn.Module] = MLP,
+        **kwargs,
+    ):
+        super().__init__()
+
+        if moments is None:
+            self.standardize = nn.Identity()
+        else:
+            mu, sigma = moments
+            self.standardize = Affine(-mu / sigma, 1 / sigma, trainable=False)
+
+        self.net = build(theta_dim + x_dim + t_dim * 2, theta_dim, **kwargs)
+
+        self.register_buffer('periods', torch.arange(t_dim) + 1)
+        self.register_buffer('zeros', torch.zeros(theta_dim))
+        self.register_buffer('ones', torch.ones(theta_dim))
+
+    def forward(self, theta: Tensor, x: Tensor, t: Tensor) -> Tensor:
+        r"""
+        Arguments:
+            theta: The parameters :math:`\theta`, with shape :math:`(*, D)`.
+            x: The observation :math:`x`, with shape :math:`(*, L)`.
+            t: The time :math:`t`, with shape :math:`(*,).`
+
+        Returns:
+            The score :math:`s_\phi(\theta, x, t)`, with shape :math:`(*,)`.
```

Comment:
To avoid future confusion, it should be highlighted here that `s_phi` is not the score, but the rescaled score.